edk2toollib/windows/capsule/inf_generator2.py

```diff
@@ -0,0 +1,399 @@
+
+        Name         - specifies the name for the INF package
+        VersionStr   - specifies the version as string in dot-tuple format (e.g. "4.15.80.0")
+        CreationDate - specifies the INF date as a string in MM/DD/YYYY format (e.g "01/01/2021")
+        Arch         - specifies the architecture as a string (e.g. "amd64")
+        Provider     - specifies the provider as a string (e.g. "Firmware Provider")
+        Manufacturer - Specifies the manufacturer as a string (e.g. "Firmware Manufacturer")
+        InfStrings   - An InfStrings object representing the "Strings" section of this INF file.
+        '''
+        self.Name = Name
+        self.VersionStr = VersionStr
+        self.Date = CreationDate
+        self.Arch = Arch
+        InfStrings.AddLocalizableString("Provider", Provider)
+        InfStrings.AddLocalizableString("MfgName", Manufacturer)
+
+    @property
+    def Name(self):
+        return self._name
+
+    @Name.setter
+    def Name(self, value):
+        # test here for invalid chars
+        if not (re.compile(r'[\w-]*$')).match(value):
+            raise ValueError("Name has invalid chars.")
+        self._name = value
+
+    @property
+    def VersionStr(self):
+        return self._versionstr
+
+    @VersionStr.setter
+    def VersionStr(self, value):
+        c = value.count(".")
+        if(c < 1) or (c > 3):
+            raise ValueError("VersionString must be in format of xx.xx -> xx.xx.xx.xx")
+        self._versionstr = value
+
+    @property
+    def Date(self):
+        return self._date
+
+    @Date.setter
+    def Date(self, value):
+        self._date = datetime.strptime(value, "%m/%d/%Y").strftime("%m/%d/%Y")
+
+    @property
+    def Arch(self):
+        return self._arch
+
+    @Arch.setter
+    def Arch(self, value):
+        key = value.lower()
+        if(key not in SUPPORTED_ARCH.keys()):
+            raise ValueError("Unsupported Architecture")
+        self._arch = SUPPORTED_ARCH[key]
+
+    def __str__(self) -> str:
+        '''Return the string representation of this InfHeader object'''
+        return textwrap.dedent(f"""\
+            ;
+            ; {self.Name}
+            ; {self.VersionStr}
+            ; Copyright (C) Microsoft Corporation. All Rights Reserved.
+            ;
+            [Version]
+            Signature="$WINDOWS NT$"
+            Class=Firmware
+            ClassGuid={{f2e7dd72-6468-4e36-b6f1-6488f42c1b52}}
+            Provider=%Provider%
+            DriverVer={self.Date},{self.VersionStr}
+            PnpLockdown=1
+            CatalogFile={self.Name}.cat
+
+            [Manufacturer]
+            %MfgName% = Firmware,NT{self.Arch}
+
+            """)
+
+
+class InfFirmware(object):
+    def __init__(self, Tag: str, Description: str, EsrtGuid: str, VersionInt: str, FirmwareFile: str,
+                 InfStrings: 'InfStrings', InfSourceFiles: 'InfSourceFiles', Rollback: bool = False,
+                 IntegrityFile: str = None) -> None:
```

Comment:
when a parameter can be None or some type it should be typed using optional which can be imported from typing.  Here is an example using optional string.  

``` py
Optional[str]
```

Reply:
Added Optional type markers.